fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/IndexKeyValueToPartialRecord.java

```diff
@@ -368,6 +368,12 @@ public void addRequiredMessageFields() {
                 if (fieldDescriptor.isRequired() && fieldDescriptor.getType() == Descriptors.FieldDescriptor.Type.MESSAGE) {
                     nestedBuilders.putIfAbsent(fieldDescriptor.getName(), new Builder(fieldDescriptor.getMessageType(), true));
                 }
+                if (fieldDescriptor.getType() == Descriptors.FieldDescriptor.Type.MESSAGE) {
+                    Builder builder = nestedBuilders.get(fieldDescriptor.getName());
```

Comment:
Would this loop be equivalent to just `nestedBuilders.values().forEach(Builder::addRequiredMessageFields)`? That may be a bit clearer (or maybe it's six of one, half dozen of the other) that effectively what's going on here is just recursively adding in builders for any required fields.